src/blocks/calendar/auth.rs

```diff
@@ -0,0 +1,303 @@
+                    let url = Url::parse(&("http://localhost".to_string() + redirect_url))
+                        .map_err(|e| CalendarError::RequestToken(e.to_string()))?;
+
+                    let (_, code_value) = url
+                        .query_pairs()
+                        .find(|pair| {
+                            let (key, _) = pair;
+                            key == "code"
+                        })
+                        .ok_or(CalendarError::RequestToken(
+                            "code query param is missing".into(),
+                        ))?;
+                    code = AuthorizationCode::new(code_value.into_owned());
+                    let (_, state_value) = url
+                        .query_pairs()
+                        .find(|pair| {
+                            let (key, _) = pair;
+                            key == "state"
+                        })
+                        .ok_or(CalendarError::RequestToken(
+                            "state query param is missing".into(),
+                        ))?;
+                    let state = CsrfToken::new(state_value.into_owned());
+                    if state.secret() != authorize_url.csrf_token.secret() {
+                        return Err(CalendarError::RequestToken(
+                            "Received state and csrf token are different".to_string(),
+                        ));
+                    }
+                }
+
+                let message = "Now your i3status-rust calendar is authorized";
+                let response = format!(
+                    "HTTP/1.1 200 OK\r\ncontent-length: {}\r\n\r\n{}",
+                    message.len(),
+                    message
+                );
+                stream.write_all(response.as_bytes())?;
+
+                return client
+                    .exchange_code(code)
+                    .set_pkce_verifier(authorize_url.pkce_code_verifier)
+                    .request(http_client)
+                    .map_err(|e| CalendarError::RequestToken(e.to_string()));
+            }
+            Err(CalendarError::TokenNotExchanged)
+        })
+        .await
+        .map_err(|e| CalendarError::RequestToken(e.to_string()))?
+    }
+}
+
+#[derive(Debug)]
+pub enum Authorize {
+    Completed,
+    AskUser(AuthorizeUrl),
+}
+
+#[derive(Debug)]
+pub struct AuthorizeUrl {
+    pkce_code_verifier: PkceCodeVerifier,
+    pub url: Url,
+    csrf_token: CsrfToken,
+}
+
+type OAuth2TokenResponse = StandardTokenResponse<EmptyExtraTokenFields, BasicTokenType>;
+
+#[derive(Debug)]
+pub struct TokenStore {
+    path: PathBuf,
+    token: Option<OAuth2TokenResponse>,
+}
+
+impl TokenStore {
+    pub fn new(path: &Path) -> Self {
+        Self {
+            path: path.into(),
+            token: None,
+        }
+    }
+
+    pub fn store(&mut self, token: OAuth2TokenResponse) -> Result<(), CalendarError> {
+        let file = File::create(&self.path).unwrap();
```

Comment:
These two unwraps can fail. Also I think we should use the tokio::fs version.

Reply:
yep, I missed those. Thx